File: rcc/communications/client/cmd/abstract_cmd_client.py
from abc import ABC
import os
import shutil

from framework.logger import Logger
from rcc.utils.shell import SHELL


class AbstractCmdClient(ABC):
    """Abstract class for commands"""

    def __init__(self, command: str):
        self._logger = Logger()
        self.__command = shutil.which(command)
        if self.__command is None:
            self._logger.error(f"Command {command} not found in path {os.environ["PATH"]}")

    @property
    def available(self):
        """Availability flag"""
        return self.__command is not None

    def _run_command(self, args="", sudo=False):
        if sudo:
            return SHELL.run_sudo_command(f"{self.__command} {args}".strip(), True)

        return SHELL.run_command(f"{self.__command} {args}".strip(), True)

File: RogPerfTuner/src/clients/shell/abstract/abstract_cmd_client.cpp
#include "clients/shell/abstract/abstract_cmd_client.hpp"

#include <cstdlib>
#include <stdexcept>

#include "utils/string_utils.hpp"

AbstractCmdClient::AbstractCmdClient(const std::string& command, const std::string& name, const bool& required)
	: Loggable(name), command_(command), available_(true) {
	if (!isCommandAvailable()) {
		if (required) {
			throw std::runtime_error("Command " + command_ + " not available");
		}

		logger->error("Command " + command_ + " not available");
		available_ = false;
	}
}

CommandResult AbstractCmdClient::run_command(const std::string& args, const bool& check, const bool& sudo) {
	if (!available_) {
		throw std::runtime_error("Command " + command_ + " not available");
	}

	std::string cmd = StringUtils::trim(command_ + " " + args);

	CommandResult result;
	if (sudo) {
		result = shell.run_elevated_command(cmd, check);
	} else {
		result = shell.run_command(cmd, check);
	}

	return result;
}

bool AbstractCmdClient::available() {
	return available_;
}

bool AbstractCmdClient::isCommandAvailable() {
	std::string cmd = "which " + command_ + " >/dev/null 2>&1";
	int ret			= std::system(cmd.c_str());
	return ret == 0;
}